lib/src/source/hosted.dart

```diff
@@ -885,41 +885,41 @@ class HostedSource extends CachedSource {
       var fileName = '$packageName-$version.tar.gz';
       var archivePath = p.join(tempDirForArchive, fileName);
 
-      // The client from `withAuthenticatedClient` will retry HTTP requests.
-      // This wrapper is one layer up and will retry checksum validation errors.
-      await retry(
+      await withAuthenticatedClient(cache, Uri.parse(description.url),
+          (client) async {
         // Attempt to download archive and validate its checksum.
-        () async {
-          final request = http.Request('GET', archiveUrl);
-          final response = await withAuthenticatedClient(cache,
-              Uri.parse(description.url), (client) => client.send(request));
-          final expectedChecksum = _parseCrc32c(response.headers, fileName);
-
-          Stream<List<int>> stream = response.stream;
-          if (expectedChecksum != null) {
-            stream = _validateStream(
-                response.stream, expectedChecksum, id, archiveUrl);
-          }
+        await client.sendWithRetries(
```

Comment:
Does this give us much over just using the retry function directly?

Reply:
TL;DR: Not really, I can change it to `retryForHttp(client, ...)` or `withHttpRetries(client, ...)`.

---

This came about from me needing access to the "request" object when trying to handle these OS errors: https://github.com/dart-lang/pub/blob/a8979a13f0a63ba28a3bb06882b367fba035aa7d/lib/src/http.dart#L363-L369

The approach we originally discussed, a `performRequest()` function argument, doesn't give us access to the "request" object.

So, I initially added a "request" argument to the HTTP retries method which was passed back to the user through a new argument in `performRequest()` like so: `performRequest(request)`. This was already starting to feel weird and there was no guarantee that the user would not use a different "request" object in `performRequest(request)`. And it was then disqualified when I saw that the "request" object cannot be re-used in retries because it gets finalized when sent.

Then, I reverted and replaced the `performRequest()` function argument with these two: `composeRequest()` and `onResponse(response)`. This lets us create a new "request" object for each attempt by invoking `composeRequest()`.

I could have kept `performRequest(request)`, but it has the same guarantee issue as above and feels super heavy to have to pass a function that just does:
```dart
(request) => await client.send(request);
```

Now, I had the choice to either pass "client" as an argument or to define the HTTP retries method as an extension on `http.Client`. I chose the extension option because `client.sendWithRetries(...)` felt nicer, but I think either option is good.

Reply:
hmm, I think we need to review the sanity of these errors.. I did some refactor to retry all of them in:
https://github.com/dart-lang/pub/pull/2254

But even then they were essentially cargo-culted :see_no_evil: 


----------------------------

But it's probably fantastic to have some logic for converting a generic `SocketException` into human readable messages like `Unable to validate SSL certificate`. That's helpful for users.

What about:
```dart
/// Program-wide limiter for concurrent network requests.
final _httpPool = Pool(16);

/// Retry [fn] when HTTP exceptions are thrown, printing [operation] in log
/// and error messages.
///
/// The [operation] should be a human readable description of the operation
/// being done. Such that messages like:
///  * "[Something went wrong] when [operation]"
///  * "Retry #2 [operation]"
Future<T> retryForHttp<T>(
  String operation,
  FutureOr<T> Function() fn,
) async {
  try {
    return await retry(
      () async => await _httpPool.withResource(() async => await fn()),
      retryIf: (e, _) =>
          e is SocketException ||
          e is TlsException ||
          e is HttpException ||
          e is PubHttpException && e.couldRetry,
      onRetry: (_, n) => log.io('Retry #${n + 1} $operation'),
      maxAttempts: math.max(
        1, // Having less than 1 attempt doesn't make sense.
        int.tryParse(Platform.environment['PUB_MAX_HTTP_RETRIES'] ?? '') ?? 7,
      ),
    );
  } on SocketException catch (e) {
    final osError = e.osError;
    if (osError == null) {
      throw PubConnectException('', couldRetry: true);
    }

    if (osError.isDnsError) {
      fail('Could not resolve URL when $operation.');
    } else if (osError.isSslError) {
      fail('Unable to validate SSL certificate when $operation.');
    }
    rethrow;
  }
}
```

Then using it becomes:
```dart
      await withAuthenticatedClient(cache, Uri.parse(description.url),
          (client) async {
        // Attempt to download archive and validate its checksum.
        await retryForHttp('downloading "$archiveUrl"', () async {
          final req = http.Request('GET', archiveUrl);
          final res = await client.send(req).timeout(Duration(seconds: 30));
          res.throwIfNotOK();

          Stream<List<int>> stream = res.stream;

          final expectedCrc32 = _parseCrc32c(res.headers, fileName);
          if (expectedCrc32 != null) {
            stream = _validateStream(stream, expectedCrc32, id, archiveUrl);
          }

          await createFileFromStream(stream, archivePath);
        });
      });
```

Reply:
I like the "operation" argument. In your version, would `_parseCrc32c` and `_validateStream` have to be changed to throw `PubHttpException` instead of `PackageIntegrityException`?

Reply:
One blocker for this approach is the usage of `withResource()`. I've tried to make that work in a previous iteration and found that it makes it very easy to create deadlocks. The `withResource()` function has many cases where the inner request is either canceled, finished, or errored but the resource doesn't get released.

This complicated piece of code ends up being crucial:
https://github.com/dart-lang/pub/blob/b0978b0e80a1a83bd422b5565d9756e07a0fc35c/lib/src/http.dart#L318-L345